### PR TITLE
fix(compiler-cli): use ModuleWithProviders type if static eval fails

### PR DESCRIPTION
**Pull request by alxhub**
When the compiler encounters a function call within an NgModule imports
section, it attempts to resolve it to an NgModule-annotated class by
looking at the function body and evaluating the statements there. This
evaluation can only understand simple functions which have a single
return statement as their body. If the function the user writes is more
complex than that, the compiler won't be able to understand it and
previously the PartialEvaluator would return a "DynamicValue" for
that import.

With this change, in the event the function body resolution fails the
PartialEvaluator will now attempt to use its foreign function resolvers to
determine the correct result from the function's type signtaure instead. If
the function is annotated with a correct ModuleWithProviders type, the
compiler will be able to understand the import without static analysis of
the function body.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts [hunk 1]
```diff
@@ -584,7 +584,7 @@ export class NgModuleDecoratorHandler implements
         throw new FatalDiagnosticError(
             ErrorCode.VALUE_HAS_WRONG_TYPE, expr,
             `Value at position ${idx} in the NgModule.${arrayName} of ${
-                className} is not a reference: ${entry}`);
+                className} is not a reference: ${entry!.constructor.name}`);
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:587]: This seems to me like a dangerous change to make, given that `entry` could legitimately be `null`/`undefined`.

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts [hunk 2]
```diff
@@ -2390,6 +2390,45 @@ runInEachFileSystem(os => {
     });
 
     describe('unwrapping ModuleWithProviders functions', () => {
+      it('should use a local ModuleWithProviders-annotated return type if a function is not statically analyzable',
+         () => {
+           env.write(`module.ts`, `
+            import {NgModule, ModuleWithProviders} from '@angular/core';
+
+            export function notStaticallyAnalyzable(): ModuleWithProviders<SomeModule> {
```
Comment (JoostK) [inline packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:2398]: How about adding a test where the return type is missing. We can then assert on the insightful diagnostic that should be reported by the FFR (i.e. [here](https://github.com/angular/angular/blob/b07b7c9f3e653408fdedba258778fd4cc8003535/packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts#L507-L511)) which would address the poor UX that we had until this change.

**Review by JoostK — CHANGES_REQUESTED** [review]
I believe this would address #36415 (FW-2048) and make #14410 obsolete.

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts (hunk 2 above)
Comment (alxhub) [reply packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:2398]: Oh, that's an interesting idea. Will do!

Comment (JoostK) [reply packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:2398]: It can still occur when someone leaves off the return type altogether.

**Review by JoostK — APPROVED** [review]

**Comment by alxhub** [comment]
[Presubmit](http://test/OCL:312572515:BASE:312572747:1590015639100:10da24d1)

**Comment by alxhub** [comment]
[Presubmit](http://test/OCL:312572515:BASE:314214782:1591050283518:7c59689f)

**Pull request closed without merging**